src/llvm-late-gc-lowering.cpp

```diff
@@ -1889,6 +1902,189 @@ void LateLowerGCFrame::PlaceGCFrameStores(State &S, unsigned MinColorRoot,
+}
+
+std::pair<SmallVector<Instruction*,4>,SmallVector<Instruction*,16>>
+LateLowerGCFrame::LocatePushPop(State &S)
+{
+    std::pair<SmallVector<Instruction*,4>,SmallVector<Instruction*,16>> res;
+    auto &push = res.first;
+    auto &pop = res.second;
+    auto global_frame = [&] {
+        push.push_back(&*(++BasicBlock::iterator(ptlsStates)));
+        for (auto &BB: *S.F) {
+            if (auto ret = dyn_cast<ReturnInst>(BB.getTerminator())) {
+                pop.push_back(ret);
+            }
+        }
+    };
+    // These are harder to handle, ignore them for now...
+    if (!S.Allocas.empty() || !S.ReturnsTwice.empty()) {
+        global_frame();
+        return res;
+    }
+    // We don't want to place any push/pop in a loop so first identify the graph of SCC nodes
+    // Each SCC node will always has the same "liveness".
+    struct SCCInfo {
+        std::vector<BasicBlock*> BBs;
+        SetVector<unsigned> Succ{};
+        SetVector<unsigned> Pred{};
+        bool HasLive = false;
+        bool MayReachLive = false;
+        SCCInfo(const std::vector<BasicBlock*> &BBs)
+            : BBs(BBs)
+        {
+        }
+    };
+    SmallVector<SCCInfo,4> sccnodes;
+    std::map<BasicBlock*,unsigned> sccmap;
+    for (auto SCCI = scc_begin(S.F); !SCCI.isAtEnd(); ++SCCI) {
+        unsigned id = sccnodes.size();
+        sccnodes.emplace_back(*SCCI);
+        auto &node = sccnodes.back();
+        for (auto bb: node.BBs) {
+            sccmap[bb] = id;
+            BBState &BBS = S.BBStates[bb];
+            node.HasLive |= BBS.HasLive;
+            for (auto succ_bb: successors(bb)) {
+                auto succ_id = sccmap[succ_bb];
+                if (succ_id < id) {
+                    node.Succ.insert(succ_id);
+                    auto &succ_node = sccnodes[succ_id];
+                    succ_node.Pred.insert(id);
+                    if (succ_node.HasLive || succ_node.MayReachLive) {
+                        node.MayReachLive = true;
+                    }
+                }
+            }
+        }
+    }
+    // We want to make sure each path through the CFG of SCC has at most one push and one pop
+    // therefore we need to mark everything that's reachable from a live node and may reach
+    // a live node has to be live too.
+    // Iterate in the forward control flow direction to do this in one pass
+    // since we've already propagated the may reach live info in the reverse direction.
+    for (auto it = sccnodes.rbegin(); it != sccnodes.rend(); ++it) {
+        if (!it->HasLive)
+            continue;
+        for (auto succ_id: it->Succ) {
+            auto &succ_node = sccnodes[succ_id];
+            if (succ_node.MayReachLive) {
+                succ_node.HasLive = true;
+            }
+        }
+    }
+    // Now the liveness is valid, mark all nodes that has all outgoing or all incoming nodes
+    // live also as live to reduce the number of edges
```

Comment:
doesn't this also increase the number of edges (unless you were to first split the edge)?